### PR TITLE
LibC: Implement some missing functions in stdio

### PR DESCRIPTION
**Pull request by SeekingBlues**

### CODE REVIEW

**Comment by sin-ack** [comment]
All the different function implementations are squished into a single function. Please separate the implementation of each function into separate commits, making sure that each commit builds on its own (see [CONTRIBUTING.md](https://github.com/SerenityOS/Serenity/blob/master/CONTRIBUTING.md)).

File: Userland/Libraries/LibC/stdio.cpp [hunk 1]
```diff
@@ -1287,19 +1325,27 @@ int vscanf(const char* fmt, va_list ap)
...
+{
+    VERIFY(filehandle);
+    ScopedFileLock::lock(filehandle);
+}
+
+int ftrylockfile(FILE* filehandle)
 {
-    dbgln("FIXME: Implement flockfile()");
+    VERIFY(filehandle);
+    return ScopedFileLock::try_lock(filehandle);
```
Comment (sin-ack) [inline Userland/Libraries/LibC/stdio.cpp:1337]: This is bogus. `ScopedFileLock` will unlock the file as soon as it is out of scope. This does not actually lock the file in anyway.

File: Userland/Libraries/LibC/stdio.cpp [hunk 2]
```diff
@@ -1159,10 +1191,16 @@ void dbgputstr(const char* characters, size_t length)
     syscall(SC_dbgputstr, characters, length);
 }
 
-char* tmpnam(char*)
+char* tmpnam(char* s)
 {
-    dbgln("FIXME: Implement tmpnam()");
-    TODO();
+    static char tmpnam_buffer[L_tmpnam];
```
Comment (sin-ack) [inline Userland/Libraries/LibC/stdio.cpp:1196]: Please use an `Array` here, in the C++ files we are not bound by C's limitations.

File: Userland/Libraries/LibC/stdio.cpp [hunk 3]
```diff
@@ -1092,9 +1114,19 @@ FILE* fopen(const char* pathname, const char* mode)
...
 {
     VERIFY(stream);
+
+    char path_buf[MAXPATHLEN];
     if (!pathname) {
-        // FIXME: Someone should probably implement this path.
-        TODO();
+        char proc_path[MAXPATHLEN];
+        if (snprintf(proc_path, sizeof(proc_path), "/proc/self/fd/%d", stream->fileno()) < 0)
+            return nullptr;
```
Comment (sin-ack) [inline Userland/Libraries/LibC/stdio.cpp:1122]: Please use `String::formatted` or `StringBuilder`.

File: Userland/Libraries/LibC/stdio.cpp [hunk 4]
```diff
@@ -31,7 +31,8 @@ struct FILE {
         : m_fd(fd)
         , m_mode(mode)
     {
-        __pthread_mutex_init(&m_mutex, nullptr);
+        __pthread_mutexattr_t mutex_attr { __PTHREAD_MUTEX_RECURSIVE };
```
Comment (sin-ack) [inline Userland/Libraries/LibC/stdio.cpp:34]: What's the reason for making the mutex recursive?

**Review by sin-ack — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibC/stdio.cpp (hunk 1 above)
Comment (SeekingBlues) [reply Userland/Libraries/LibC/stdio.cpp:1337]: I added static functions to `ScopedFileLock` that only lock or unlock the file. The name might be misleading, but we should keep the lock functions of `FILE` private and not add any friend functions.

File: Userland/Libraries/LibC/stdio.cpp (hunk 4 above)
Comment (SeekingBlues) [reply Userland/Libraries/LibC/stdio.cpp:34]: POSIX [says](https://pubs.opengroup.org/onlinepubs/9699919799/functions/ftrylockfile.html) that `FILE` should maintain a lock count. So if this is to be implemented in a mutex, it should be recursive.

File: Userland/Libraries/LibC/stdio.cpp (hunk 2 above)
Comment (SeekingBlues) [reply Userland/Libraries/LibC/stdio.cpp:1196]: Okay. I was hesitant about using these things in stdio, but I will change them right away.

File: Userland/Libraries/LibC/stdio.cpp (hunk 3 above)
Comment (SeekingBlues) [reply Userland/Libraries/LibC/stdio.cpp:1122]: No problem.

File: Userland/Libraries/LibC/stdio.cpp (hunk 1 above)
Comment (sin-ack) [reply Userland/Libraries/LibC/stdio.cpp:1337]: Then please change that. `ScopedFileLock` locks a file within a scope. Please don't add things that allow it to do things outside a scope.

**Comment by SeekingBlues** [comment]
Everything should be fixed now. Thanks for the feedback :^)

**Pull request closed without merging**